test/implementation-report/EarlFormatter.cpp

```diff
@@ -0,0 +1,116 @@
+#include <iostream>
+#include <ctime>
+#include "EarlFormatter.h"
+
+EarlFormatter::EarlFormatter()
+{
+    // insert initial namespace
+    RdfNamespace ns;
+    ns.uri = "http://www.w3.org/ns/earl#";
+    ns.prefix = "earl";
+    namespaces.insert( ns );
+}
+
+void EarlFormatter::format( std::stringstream& ss, RdfData* data, int depth )
+{
+    // check if RdfData has a subject, if not then this is the subject and
+    // needs to be encapsulated in square brackets
+
+    // get the main object
+    auto& obj = data->subject;
+    // is the object empty? If so begin an array
+    if ( obj.name == "" )
+    {
+        ss << "[ ";
+        depth++;
+    }
+    else
+    {
+        // check for a namespace
+        if ( obj.ns.prefix != "" )
+        {
+            addNamespace( obj.ns );
+            ss << obj.ns.prefix << ':' << obj.name;
+        } else {
+            // check for last element in data
+            // TODO: better way of handling "a"
+            if ( data->objects.empty() && obj.name != "a" )
```

Comment:
I have tried to create a RDF data structure in where we have an RDFNamespace that contains the URI for use in the context and a prefix that is used in the main RDF data:
```
prefix: [
  { "dc":"http://purl.org/dc/terms/" }
  ...
```

RDFObject then contains a RDFNamespace and name.  These are basically key -> value properties where the RDFNamespace describes the Object and the Name is the Property of the Object.

RDFData then takes an RDFObject that is the Subject, then relates other RDFData in order to build up a tree of data thats used to build the output:

```
[ github:hawkmauk [
  a [ foaf:person: -  ] ] ;
  foaf:name: "Michael Hawkins" ;
]
```

'a' can be used to nest properties in the case of assertions which makes it different to when a 'normal' RDFObject is used:

```
[ a [ earl:assertion ] ;
  earl:subject [ doap:name "jsonld-cpp" ];
  ...
  [ a [ earl:TestResult ] ;
    earl:outcome [ earl:passed ] ;
  ]
]
```
This TODO was a prompt to see if there was a better way of handling this as I ike to avoid special cases if possible.  Might not be worth opening a can of worms with this?

Reply:
Yes, seems like a good workaround for this case, and since it is an internal-only detail and not something emitted into the turtle output, it is probably fine to leave it as is if it is working fine.